File: app/src/main/java/com/elementary/tasks/core/view_models/google_tasks/GoogleTaskListViewModel.java
package com.elementary.tasks.core.view_models.google_tasks;

import android.app.Application;

import com.elementary.tasks.core.cloud.Google;
import com.elementary.tasks.core.data.models.GoogleTask;
import com.elementary.tasks.core.data.models.GoogleTaskList;
import com.elementary.tasks.core.utils.SuperUtil;
import com.elementary.tasks.core.view_models.Commands;

import java.io.IOException;
import java.util.List;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.lifecycle.LiveData;
import androidx.lifecycle.ViewModel;
import androidx.lifecycle.ViewModelProvider;

/**
 * Copyright 2018 Nazar Suhovich
 * <p/>
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * <p/>
 * http://www.apache.org/licenses/LICENSE-2.0
 * <p/>
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
public class GoogleTaskListViewModel extends BaseTaskListsViewModel {

    public LiveData<GoogleTaskList> googleTaskList;
    public LiveData<GoogleTaskList> defaultTaskList;
    public LiveData<List<GoogleTask>> googleTasks;

    public GoogleTaskListViewModel(Application application, @Nullable String listId) {
        super(application);
        googleTaskList = getAppDb().googleTaskListsDao().loadById(listId);
        defaultTaskList = getAppDb().googleTaskListsDao().loadDefault();
        if (listId == null) {
            googleTasks = getAppDb().googleTasksDao().loadAll();
        } else {
            googleTasks = getAppDb().googleTasksDao().loadAllByList(listId);
        }
    }

    public void newGoogleTaskList(@NonNull GoogleTaskList googleTaskList) {
        Google google = Google.getInstance(getApplication());
        if (google == null || google.getTasks() == null) {
            return;
        }
        boolean isConnected = SuperUtil.isConnected(getApplication());
        if (!isConnected) {
            result.postValue(Commands.FAILED);
            return;
        }
        isInProgress.postValue(true);
        run(() -> {
            google.getTasks().insertTasksList(googleTaskList.getTitle(), googleTaskList.getColor());
            end(() -> {
                isInProgress.postValue(false);
                result.postValue(Commands.SAVED);
            });
        });
    }

    public void updateGoogleTaskList(@NonNull GoogleTaskList googleTaskList) {
        Google google = Google.getInstance(getApplication());
        if (google == null || google.getTasks() == null) {
            return;
        }
        boolean isConnected = SuperUtil.isConnected(getApplication());
        if (!isConnected) {
            result.postValue(Commands.FAILED);
            return;
        }
        isInProgress.postValue(true);
        run(() -> {
            getAppDb().googleTaskListsDao().insert(googleTaskList);
            try {
                google.getTasks().updateTasksList(googleTaskList.getTitle(), googleTaskList.getListId());
                end(() -> {
                    isInProgress.postValue(false);
                    result.postValue(Commands.SAVED);
                });
            } catch (IOException e) {
                end(() -> {
                    isInProgress.postValue(false);
                    result.postValue(Commands.FAILED);
                });
            }
        });
    }

    public void saveLocalGoogleTaskList(@NonNull GoogleTaskList googleTaskList) {
        isInProgress.postValue(true);
        run(() -> {
            getAppDb().googleTaskListsDao().insert(googleTaskList);
            end(() -> {
                isInProgress.postValue(false);
                result.postValue(Commands.SAVED);
            });
        });
    }

    public static class Factory extends ViewModelProvider.NewInstanceFactory {

        private Application application;
        @Nullable
        private String id;

        public Factory(Application application, @Nullable String id) {
            this.application = application;
            this.id = id;
        }

        @NonNull
        @SuppressWarnings("unchecked")
        @Override
        public <T extends ViewModel> T create(@NonNull Class<T> modelClass) {
            return (T) new GoogleTaskListViewModel(application, id);
        }
    }
}

File: app/src/main/java/com/elementary/tasks/core/view_models/google_tasks/GoogleTaskListViewModel.kt
package com.elementary.tasks.core.view_models.google_tasks

import androidx.lifecycle.LiveData
import com.elementary.tasks.core.app_widgets.UpdatesHelper
import com.elementary.tasks.core.cloud.GTasks
import com.elementary.tasks.core.data.dao.GoogleTaskListsDao
import com.elementary.tasks.core.data.dao.GoogleTasksDao
import com.elementary.tasks.core.data.models.GoogleTask
import com.elementary.tasks.core.data.models.GoogleTaskList
import com.elementary.tasks.core.utils.launchDefault
import com.elementary.tasks.core.utils.withUIContext
import com.elementary.tasks.core.view_models.Commands
import com.elementary.tasks.core.view_models.DispatcherProvider
import com.google.api.services.tasks.model.TaskLists
import timber.log.Timber
import java.io.IOException
import java.util.Random

class GoogleTaskListViewModel(
  listId: String,
  gTasks: GTasks,
  dispatcherProvider: DispatcherProvider,
  updatesHelper: UpdatesHelper,
  googleTasksDao: GoogleTasksDao,
  googleTaskListsDao: GoogleTaskListsDao
) : BaseTaskListsViewModel(
  gTasks,
  dispatcherProvider,
  updatesHelper,
  googleTasksDao,
  googleTaskListsDao
) {

  var googleTaskList: LiveData<GoogleTaskList>
  val defaultTaskList = googleTaskListsDao.loadDefault()
  var googleTasks: LiveData<List<GoogleTask>>
  private var isSyncing = false

  init {
    Timber.d("GoogleTaskListViewModel: $listId")
    googleTaskList = googleTaskListsDao.loadById(listId)
    googleTasks = googleTasksDao.loadAllByList(listId)
  }

  fun sync() {
    if (!gTasks.isLogged) {
      postCommand(Commands.FAILED)
      return
    }
    if (isSyncing) return
    isSyncing = true
    postInProgress(true)
    launchDefault {
      var lists: TaskLists? = null
      try {
        lists = gTasks.taskLists()
      } catch (e: IOException) {
        e.printStackTrace()
      }

      if (lists != null && lists.size > 0 && lists.items != null) {
        for (item in lists.items) {
          val listId = item.id
          var taskList = googleTaskListsDao.getById(listId)
          if (taskList != null) {
            taskList.update(item)
          } else {
            val r = Random()
            val color = r.nextInt(15)
            taskList = GoogleTaskList(item, color)
          }
          googleTaskListsDao.insert(taskList)
          val tasks = gTasks.getTasks(listId)
          if (tasks.isEmpty()) {
            withUIContext {
              postInProgress(false)
              postCommand(Commands.UPDATED)
              updatesHelper.updateTasksWidget()
            }
          } else {
            val googleTasks = ArrayList<GoogleTask>()
            for (task in tasks) {
              var googleTask = googleTasksDao.getById(task.id)
              if (googleTask != null) {
                googleTask.listId = listId
                googleTask.update(task)
              } else {
                googleTask = GoogleTask(task, listId)
              }
              googleTasks.add(googleTask)
            }
            googleTasksDao.insertAll(googleTasks)
            withUIContext {
              postInProgress(false)
              postCommand(Commands.UPDATED)
              updatesHelper.updateTasksWidget()
            }
          }
        }
      }
      isSyncing = false
    }
  }

  fun newGoogleTaskList(googleTaskList: GoogleTaskList) {
    if (!gTasks.isLogged) {
      postCommand(Commands.FAILED)
      return
    }
    postInProgress(true)
    launchDefault {
      gTasks.insertTasksList(googleTaskList.title, googleTaskList.color)
      postInProgress(false)
      postCommand(Commands.SAVED)
    }
  }

  fun updateGoogleTaskList(googleTaskList: GoogleTaskList) {
    if (!gTasks.isLogged) {
      postCommand(Commands.FAILED)
      return
    }
    postInProgress(true)
    launchDefault {
      googleTaskListsDao.insert(googleTaskList)
      try {
        gTasks.updateTasksList(googleTaskList.title, googleTaskList.listId)
        postInProgress(false)
        postCommand(Commands.SAVED)
      } catch (e: IOException) {
        postInProgress(false)
        postCommand(Commands.FAILED)
      }
    }
  }

  fun clearList(googleTaskList: GoogleTaskList) {
    if (!gTasks.isLogged) {
      postCommand(Commands.FAILED)
      return
    }
    postInProgress(true)
    launchDefault {
      val googleTasks = googleTasksDao.getAllByList(googleTaskList.listId, GTasks.TASKS_COMPLETE)
      googleTasksDao.deleteAll(googleTasks)
      gTasks.clearTaskList(googleTaskList.listId)
      postInProgress(false)
      postCommand(Commands.UPDATED)
      withUIContext {
        updatesHelper.updateTasksWidget()
      }
    }
  }

  fun saveLocalGoogleTaskList(googleTaskList: GoogleTaskList) {
    postInProgress(true)
    launchDefault {
      googleTaskListsDao.insert(googleTaskList)
      postInProgress(false)
      postCommand(Commands.SAVED)
    }
  }
}
